File: fsharp/ch2/retirement/Input.fs
module Retirement.Input

open System

type UserInput =
    | Empty
    | Success of int
    | Failure
    | Negative

let private makeParseState (success:bool,result:int) =
    match success with
    | true -> if result < 0 then Negative else Success(result)
    | false -> Failure

let parse (str:string) =
    Int32.TryParse str
    |> makeParseState

File: clojure/ch2/retirement/src/brianium/retirement/input.clj
(ns brianium.retirement.input
  (:require [clojure.spec.alpha :as s]
            [brianium.retirement.input.spec :as input]))

(defn make-parse-state
  [i]
  (if (< i 0)
    {:state :negative :value i}
    {:state :success :value i}))

(s/fdef make-parse-state
  :args (s/cat :i int?)
  :ret  ::input/state)

(defn parse
  [s]
  (try
    (make-parse-state (Integer/parseInt s))
    (catch Exception _
      {:state :failure})))

(s/fdef parse
  :args (s/cat :s string?)
  :ret  (s/tuple int? boolean?))

(defn success?
  [parse-state]
  (= :success (:state parse-state)))

(s/fdef success?
  :args (s/cat :state ::input/state)
  :ret  boolean?)

(def feedback {:failure  "Value must be a number."
               :negative "Value cannot be negative."
               :empty    "Value cannot be empty."})
